agent/consul/internal_endpoint.go

```diff
@@ -123,6 +123,21 @@ func (m *Internal) ServiceDump(args *structs.ServiceDumpRequest, reply *structs.
 			}
 
 			reply.Nodes = raw.(structs.CheckServiceNodes)
+			for i, n := range reply.Nodes {
+				if n.Service.Kind == structs.ServiceKindTypical {
+					svcName := n.Service.CompoundServiceName()
+					idx, gatewayServices, err := state.GatewayServicesForService(ws, svcName.Name, &svcName.EnterpriseMeta)
+					if err != nil {
+						return err
+					}
+
+					if idx > reply.Index {
+						reply.Index = idx
+					}
+					reply.Nodes[i].GatewayServices = gatewayServices
+				}
+			}
```

Comment:
This section of code has the effect that we are currently ignoring "virtual" services in our final `AssociatedServiceCount` value, since we only ask for services that exist in the output of `ServiceDump`.

The other way I tried to do this was by attaching all of the `GatewayServices` to the gateway's `CheckServiceNode`, but that had the effect of where `CheckServiceNode` and `ServiceInfo` are the exact same struct, but the `GatewayService` field was used differently between them.

That is, in the `GatewayServiceDump` RPC, we created `ServiceInfo` structs for every service and attach the associated `GatewayService` to it. We maintain that consistency here, by looking up `GatewayServicesForService` and attaching that to the service's `CheckServiceNode`.